### PR TITLE
Skin OpenMC DAGMC geometries

### PR DESCRIPTION
**Pull request by aprilnovak**
This adds on-the-fly skinning of OpenMC DAGMC geometries according to contours in temperature based on the implementation in [Aurora](https://github.com/aurora-multiphysics/aurora). The elements are grouped into "bins" according to temperature, material, and density, to re-generate the DAGMC cells on the fly. For example, with a starting DAGMC geometry with 2 cells (each with a unique material), like so:

Screen Shot 2023-02-22 at 4 03 14 PM

The skinner will re-generate the geometry by binning elements. The other parts of `OpenMCCellAverageProblem` then treat the new cells as usual, and send volume-averaged temperature and density into them. For more info, see [new documentation](https://cardinal.cels.anl.gov/source/problems/OpenMCCellAverageProblem.html#:~:text=cad%20geometry%20skinning).

Screen Shot 2023-02-22 at 4 40 41 PM

## Differences from Aurora

Aside from the differences in the skinning operation itself (see https://github.com/neams-th-coe/cardinal/pull/564), there are not many fundamental design differences from Aurora. But it is worth summarizing them:

- Instead of setting the new cell temperatures and densities to the bin "midpoint," we directly compute the volume-averaged temperature/density from the elements which map to each bin. 

## How did I test this

The regression tests do a 1:1 comparison between the skinned DAGMC models and equivalent CSG versions. For example, in the image above, an initial OpenMC model with 2 cells is re-skinned into 8 new cells at different temperatures. I built an identical CSG version of this same 8-cell geometry (with the same temperature values in each cell as in the skinned model). I was able to match exactly the values for mesh tally bins and k-eigenvalue, which proves equivalent behavior.

Other tests in this PR are error checking.

##  Limitations

There are a number of limitations in the present implementation that will be addressed in future PRs because those developments either (i) require other open PRs to be merged, or (ii) are good motivation to change some of the design in OpenMC itself. I also wanted to keep the PR smaller for an easier review. The current implementation:

- does not allow density binning. With density binning, we need to add new OpenMC materials after the first time step, because the notion of "density" is attached to a material (as opposed to "temperature," which can be modified for a cell based on the `Cell` object itself, OR the `Material` filling the cell). I'd prefer to just change OpenMC to allow density to be applied to the `Cell` object directly. I think technically right now this means that the skinning operation can only be applied to solid-only OpenMC models, because otherwise the density transfer will complain about a material being duplicated across multiple fluid cells. So this will definitely be addressed in some manner in the future.
- does not support UWUW; I'm not sure how widely UWUW is used, but I wanted to keep the implementation simple for now

### CODE REVIEW

File: doc/content/source/problems/OpenMCCellAverageProblem.md
Comment (pshriwise) [inline doc/content/source/problems/OpenMCCellAverageProblem.md:417]: We can do all or some of these?

```suggestion
and/or subdomain (mesh block). A new DAGMC cell will then be created by "skinning" the mesh
```

File: doc/content/source/problems/OpenMCCellAverageProblem.md
Comment (pshriwise) [inline doc/content/source/problems/OpenMCCellAverageProblem.md:453]:
```suggestion
A "graveyard" volume is automatically built around the new DAGMC model between two bounding boxes. This
```

File: doc/content/source/problems/OpenMCCellAverageProblem.md
Comment (aprilnovak) [reply doc/content/source/problems/OpenMCCellAverageProblem.md:417]: In the long run, we'll be able to do all of these simultaneously. I didn't implement the necessary material-reinits for density binning, so `OpenMCCellAverageProblem` currently throws an error if trying to use density binning (just to protect any user who picks this up right away)

**Comment by aprilnovak** [comment]
@pshriwise @helen-brooks this is ready for review now. Since #599 was merged last week, I've added support for cell tallies (where each DAGMC cell is added to a cell instance filter) and verified again with the same exact-CSG case. This just provides extra capability and confirms that the implementation is correct.

**Review by helen-brooks — COMMENTED** [review]
This is great. I don't think I have any criticisms, or requests for changes, just one question. I can't see where the temperature gets set on the OpenMC cells. Could you point me at the relevant lines?

**Comment by aprilnovak** [comment]
> I can't see where the temperature gets set on the OpenMC cells. Could you point me at the relevant lines?

Summarizing here the outcome of an in-person discussion:

The temperature gets set within other logic (untouched by this PR) in `OpenMCCellAverageProblem`. The code loops over the elements in the `[Mesh]` and forms a map b/w cells <--> element IDs. It then sets the temperature in each cell as a volume average over all that cell's corresponding elements.

File: include/userobjects/MoabSkinner.h
Comment (pshriwise) [inline include/userobjects/MoabSkinner.h:127]: For my understanding, the case in which we don't want the skinner to build a graveyard is if the boundary conditions are picked up from the MOOSE mesh?

File: src/userobjects/MoabSkinner.C
Comment (pshriwise) [inline src/userobjects/MoabSkinner.C:591]: I think variables for temperature and density tolerances would be good here in place of the literals here. IMO they don't necessarily need to be exposed as user parameters (yet), but giving them names will help with clarity a bit.

**Review by pshriwise — CHANGES_REQUESTED** [review]
This looks really good @aprilnovak! Excited to get this functionality merged in. A few minor comments/questions from me here.

File: doc/content/source/problems/OpenMCCellAverageProblem.md
Comment (aprilnovak) [reply doc/content/source/problems/OpenMCCellAverageProblem.md:491]: Correct, because we're not setting any reflective BCs in the skinned OpenMC model, it wouldn't be compatible with all the symmetry mapping stuff. I added an error if someone tries to use it.

File: include/base/OpenMCCellAverageProblem.h
Comment (aprilnovak) [reply include/base/OpenMCCellAverageProblem.h:949]: I think I did this so that when I later iterate through `_subdomain_to_material` that the print statements go through the subdomain IDs in order. But I'm not sure I actually tested it with `std::unordered_set` to see what the behavior was.

File: include/userobjects/MoabSkinner.h
Comment (aprilnovak) [reply include/userobjects/MoabSkinner.h:127]: For the current implementation, we always want the skinner to build a graveyard. This function is used to detect if the user set `build_graveyard` on the `MoabSkinner`. If the user forgot it, then instead of erroring out we set the graveyard flag to true from inside `OpenMCCellAverageProblem`. So basically, this is just being more forgiving if the user forgot to turn the graveyard feature on, we do it for them under the hood.

File: src/base/OpenMCCellAverageProblem.C
Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:613]: Oh nice, I didn't know about that one. Can I use `openmc::DAGMC_ENABLED` everywhere that `ifdef ENABLE_DAGMC` is used though? For instance, in the constructor there is:

```
    _symmetry(nullptr),
#ifdef ENABLE_DAGMC
    _skinner(nullptr),
#endif
```

I feel like that wouldn't work if I swapped them.

File: src/base/OpenMCCellAverageProblem.C
Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:2800]: I honestly don't know 100% what these lines are doing. They come from Aurora; I initially tried it without the lines, and I think the code seg faulted (but was fixed once adding them in). @helen-brooks would know more

File: src/userobjects/MoabSkinner.C
Comment (aprilnovak) [reply src/userobjects/MoabSkinner.C:591]: Agreed, good call

File: src/base/OpenMCCellAverageProblem.C
Comment (pshriwise) [reply src/base/OpenMCCellAverageProblem.C:613]: True ture. What we ended up doing in OpenMC was to wrap the entire C++ and header file in a `DAGMC` `#ifdef`. That kept them mostly contained the `dagmc.cpp` and `dagmc.h` files. They can still be included elsewhere without `#ifdef`'s when DAGMC is not enabled, but no functionality is added. We could do the same here in `MOABSkinner.C/H`

File: include/userobjects/MoabSkinner.h
Comment (pshriwise) [reply include/userobjects/MoabSkinner.h:127]: Got it. Then I guess maybe my question above relates back to Aurora and is more of a @helen-brooks question.

File: include/base/OpenMCCellAverageProblem.h
Comment (pshriwise) [reply include/base/OpenMCCellAverageProblem.h:949]: Sounds good. I've got no preference myself in this case. Thanks!

File: src/base/OpenMCCellAverageProblem.C
Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:613]: Sorry I'm not sure I follow what is being suggested - could you please rephrase?

File: include/userobjects/MoabSkinner.h
Comment (aprilnovak) [reply include/userobjects/MoabSkinner.h:127]: Right, I think longer term we would not want the graveyard if MOOSE is setting up reflective/periodic/etc. BCs for us.

File: src/base/OpenMCCellAverageProblem.C
Comment (pshriwise) [reply src/base/OpenMCCellAverageProblem.C:613]: Something like along these lines

MoabSkinner.H:

```cpp
// other includes

#ifdef DAGMC_ENABLED
#include "moab/Core.hpp"
#include "moab/Skinner.hpp"
#include "moab/GeomTopoTool.hpp"
#include "MBTagConventions.hpp"
#endif

class MoabSkinner : public GeneralUserObject {
public:
  // Constructors
  MoabSkinner(const InputParameters & parameters);

  // MOOSE methods
  static InputParameters validParams();

  virtual void execute() override;

  virtual void initialize() override;

  virtual void finalize() override;

  virtual void threadJoin(const UserObject & /* uo */) override {}


#ifdef DAGMC_ENABLED
  // MOAB methods and data members (anything from the MOAB namespace)
  cosnt std::shared_ptr<moab::Interface>& moabPtr() { return _moab; }
#endif

```

MoabSkinner.C:

```cpp
#ifndef DAGMC_ENABLED
void MoabSkinner::execute() { mooseError("DAGMC not enabled"); }

// other MOOSE methods defined here
#else
void
MoabSkinner::execute()
{
  if (_standalone)
    update();
}

// other MOAB-based method definitions here

#endif
```

Comment (pshriwise) [reply src/base/OpenMCCellAverageProblem.C:613]: The idea is that we'd put enough functionality in the blocks where `DAGMC_ENBALED` isn't defined that we can still include it in other files without needing to keep track of where to put new `#ifdef`'s when using the `_skinner` object in the `OpenMCCellAverageProblem` source code.

File: include/userobjects/MoabSkinner.h
Comment (pshriwise) [reply include/userobjects/MoabSkinner.h:127]: Right on. Not something to hold up this PR over by any means. I can find out more later 👍🏻

File: src/base/OpenMCCellAverageProblem.C
Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:613]: Gotcha, thanks for going into more detail. The only downside is that the test suite currently is relying on `MoabSkinner` not existing altogether when the DAGMC build is disabled, to know when to skip those tests. So if we go this route, I'll need to re-add a dummy empty object (like this `DagMCPostprocessor` I deleted earlier: https://github.com/neams-th-coe/cardinal/pull/606).

I don't have a strong opinion either way - which do you prefer?

Comment (pshriwise) [reply src/base/OpenMCCellAverageProblem.C:613]: Can we also use `openmc::DAGMC_ENABLED` varaible in the test quite to skip those tests?

Even if not, I do prefer this approach because it contains those preprocessor statements more cleanly and makes external code easier understand, but I'm happy to defer to you on this in the end!

Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:613]: No, unfortunately not. It's not C++ code, it's syntax special to CIVET. Looks like this:

```
[Tests]
  [test]
    type = Exodiff
    ...
    required_objects = 'MoabSkinner'
  []
[]
```

I like your design idea though, I do feel like it is cleaner and if we design correctly less prone to use forgetting about an ifdef. I'll update to try it out

Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:613]: Actually the more I think about it, I'm not sure this is a good idea. The downsides I see:

- Need to add the dummy `DagMCPostprocessor` back in
- `MoabSkinner` will now always exist, and nothing stops the user from adding it in their input files or seeing it build with the documentation (which could be confusing)
- More code fragmentation in `MoabSkinner` to have valid "routes" for whether DAGMC is enabled/disabled.

I think I can use `openmc::DAGMC_ENABLED` in `OpenMCCellAverageProblem` in a few places to replace the ifdefs though, as well as moving more code outside of the ifdefs, so I will do that to minimize the amount of lines the ifdefs are enclosing.

**Comment by aprilnovak** [comment]
I just pushed a new commit that adds support for non-unity scaling. I'm not exactly sure what was happening before, but I think I had a mistake in my input file somehow (very weird that it worked for powers-of-2 scaling, but idk).

I basically started from scratch with a totally different imposed temperature distribution (I thought maybe there was roundoff errors in my case), and am able to one-for-one match with different values for scaling.

File: src/base/OpenMCCellAverageProblem.C
Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C]: Actually, I don't think `openmc::DAGMC_ENABLED` is going to be viable on CIVET - it seems like we're configuring to just build OpenMC once with DAGMC enabled, and then separately testing with/without DAGMC in Cardinal. I'd rather not go through the steps to build OpenMC 2x if OpenMC' test suite is ultimately responsible for ensuring correct DAGMC behavior _inside of_ OpenMC.

Comment (pshriwise) [reply src/base/OpenMCCellAverageProblem.C:606]: Thanks for trying this out, @aprilnovak! Much appreciated. Maybe we can figure out a way to do it in the future or refactor DAGMC-relevant code into an intermediate layer.

> Need to add the dummy DagMCPostprocessor back in

Agreed that the dummy `DAGMCPostProcessor` just for the sake of testing is awkward. 😖 

> MoabSkinner will now always exist, and nothing stops the user from adding it in their input files or seeing it build with the documentation (which could be confusing)

A user could have a `MoabSkinner` object in their input file with either approach and end up with a failed sim, yeah? If anything, it seems like our own implementations of those inherited `GeneralUserObject` methods and/or constructors allows us to provide a more specific error message than the one that would be provided by default for attempting to use a non-existing user object. Am I off on this?

> More code fragmentation in MoabSkinner to have valid "routes" for whether DAGMC is enabled/disabled.

Once we've provided implementations for the inherited methods of the `GeneralUserObject` for the case without DAGMC, would we really have to add any others in the future? (aside from the case where the `GeneralUserObject` interface changes).

> Further, I don't think openmc::DAGMC_ENABLED is going to be viable on CIVET - it seems like we're configuring to just build OpenMC once with DAGMC enabled, and then separately testing with/without DAGMC in Cardinal. I'd rather not go through the steps to build OpenMC 2x if OpenMC' test suite is ultimately responsible for ensuring correct DAGMC behavior inside of OpenMC.

I can sympathize with that. Having to build multiple versions of dependencies in CI isn't ideal.

Comment (aprilnovak) [reply src/base/OpenMCCellAverageProblem.C:606]: I do like the design idea, but I'd say we revisit in the future. Mostly because of the needing-to-build-OpenMC-2x on CIVET, and:

> own implementations of those inherited GeneralUserObject methods and/or constructors allows us to provide a more specific error message than the one that would be provided by default for attempting to use a non-existing user object.

True, if someone forgets to build with DAGMC and tries using an input file with `MoabSkinner`, they get an error like `object MoabSkinner not found in library!`. But currently in Cardinal basically every object can trigger an error like that in some way or another, because most objects are inside include guards for `ENABLE_NEKRS`, `ENABLE_OPENMC`, and `ENABLE_DAGMC`. It would be nice to give a more instructive error message for all of these objects, but that may take more effort to add an insertion point in the MOOSE framework wherever that error is thrown (it's not in the constructor of the object).

**Review by pshriwise — APPROVED** [review]
Looks good @aprilnovak! Excited to try this out.

**Pull request merged**
